common/persistence/serialization/blob.go

```diff
@@ -223,16 +224,83 @@ func ChecksumToBlob(checksum *persistencespb.Checksum) (commonpb.DataBlob, error
 	return proto3Encode(checksum)
 }
 
-func ChecksumFromBlob(b []byte, proto string) (*persistencespb.Checksum, error) {
+func ChecksumFromBlob(binary []byte, encoding string) (*persistencespb.Checksum, error) {
 	result := &persistencespb.Checksum{}
-	err := proto3Decode(b, proto, result)
+	err := proto3Decode(binary, encoding, result)
 	if err != nil || result.GetFlavor() == enumsspb.CHECKSUM_FLAVOR_UNSPECIFIED {
 		// If result is an empty struct (Flavor is unspecified), replace it with nil, because everywhere in the code checksum is pointer type.
 		return nil, err
 	}
 	return result, nil
 }
 
+func QueueMessageToBlob(message *persistencespb.QueueMessage) (commonpb.DataBlob, error) {
+	// TODO change ENCODING_TYPE_JSON to ENCODING_TYPE_PROTO3
+	return encode(message, enumspb.ENCODING_TYPE_JSON)
+}
+
+func QueueMessageFromBlob(binary []byte, encoding string) (*persistencespb.QueueMessage, error) {
+	result := &persistencespb.QueueMessage{}
+	return result, decode(binary, encoding, result)
+}
+
+func QueueMetadataToBlob(metadata *persistencespb.QueueMetadata) (commonpb.DataBlob, error) {
+	// TODO change ENCODING_TYPE_JSON to ENCODING_TYPE_PROTO3
+	return encode(metadata, enumspb.ENCODING_TYPE_JSON)
+}
+
+func QueueMetadataFromBlob(binary []byte, encoding string) (*persistencespb.QueueMetadata, error) {
+	result := &persistencespb.QueueMetadata{}
+	return result, decode(binary, encoding, result)
+}
+
+func encode(
+	object interface{},
+	encoding enumspb.EncodingType,
+) (commonpb.DataBlob, error) {
+	if object == nil {
+		return commonpb.DataBlob{
+			Data:         nil,
+			EncodingType: encoding,
+		}, nil
+	}
+
+	switch encoding {
+	case enumspb.ENCODING_TYPE_JSON:
```

Comment:
I am surprised to see JSON here.

Reply:
me too: 
https://github.com/temporalio/temporal/blob/b70271b98d49927953c6916a56ceac573002916c/common/persistence/sql/sqlplugin/mysql/queue.go#L93